src/actions.cpp

```diff
@@ -219,13 +220,19 @@ Receipt evm_contract::execute_tx( eosio::name miner, Block& block, Transaction&
         inevm.emplace(get_self(), get_self().value);
     };
 
+    bool is_special_signature = silkworm::is_special_signature(tx.r, tx.s);
+
     tx.from.reset();
     tx.recover_sender();
     eosio::check(tx.from.has_value(), "unable to recover sender");
     LOGTIME("EVM RECOVER SENDER");
 
-    if(from_self) {
-        check(is_reserved_address(*tx.from), "actions from self without a reserved from address are unexpected");
+    // 1 Reject special signature NOT from self.
+    // 2 All other cases (normal signature or special signature from self) will be accepted.
+    // 3 If the from address is reserved address, the special balance needs some process.
+    check(from_self || !is_special_signature, "bridge signature used outside of bridge transaction");
+    
+    if (is_reserved_address(*tx.from)) {
```

Comment:
Use `is_special_signature && is_reserved_address(*tx.from)` so we don't have to pay the computation cost of unnecessarily evaluating the `is_reserved_address` function for all the regular EVM transactions that do not have a special signature.